### PR TITLE
SAML Implementation in Kratos

### PR DESCRIPTION
**Pull request by ThibHrrd**
# SAML Implementation 

## Related issue(s)

https://github.com/ory/kratos/issues/1928

### CODE REVIEW

File: embedx/config.schema.json [hunk 1]
```diff
@@ -312,6 +312,225 @@
...
+          ]
+        },
+        "label": {
+          "title": "Optional string which will be used when generating labels for UI buttons.",
+          "type": "string",
+          "examples": [
+            "Microsoft Active Directory"
+          ]
+        },
+        "public_cert_path": {
```
Comment (psauvage0) [inline embedx/config.schema.json:383]: certificate_path ? Or public_key_certificate_path ?

File: embedx/config.schema.json [hunk 2]
```diff
@@ -312,6 +312,225 @@
...
+                  },
+                  "idp_sso_url":{
+                    "const":{}
+                  }
+                }
+              }
+            }
+          ]
+        },
+        "attributes_map": {
```
Comment (psauvage0) [inline embedx/config.schema.json:493]: The attributes mapping is given in the config ? What if I have an attribute that is not provided here ? For example `group` ?

File: internal/httpclient/api/openapi.yaml [hunk 3]
```diff
@@ -1185,6 +1185,51 @@ paths:
       summary: Create a Logout URL for Browsers
       tags:
       - v0alpha2
+  /self-service/methods/saml/auth:
```
Comment (psauvage0) [inline internal/httpclient/api/openapi.yaml]: Not related to the doc but to the route of this call (I don't know where the route is defined). 

In the [Ory Documentation for the API](https://www.ory.sh/docs/kratos/reference/api#operation/initializeSelfServiceLoginFlowWithoutBrowser), we can see that when there are different calls to make for browser vs API, they usually end the route with `/browser` or `/api`.  Maybe the same can be applied here ?

File: selfservice/flow/saml/handler.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+	var c samlstrategy.ConfigurationCollection
+	conf := config.SelfServiceStrategy("saml").Config
+	if err := jsonx.
+		NewStrictDecoder(bytes.NewBuffer(conf)).
+		Decode(&c); err != nil {
+		return errors.Wrapf(err, "Unable to decode config %v", string(conf))
+	}
+
+	// Key pair to encrypt and sign SAML requests
+	keyPair, err := tls.LoadX509KeyPair(strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PublicCertPath, "file://", "", 1), strings.Replace(c.SAMLProviders[len(c.SAMLProviders)-1].PrivateKeyPath, "file://", "", 1))
```
Comment (psauvage0) [inline selfservice/flow/saml/handler.go]: No Out Of Bound possible here ?

File: selfservice/flow/saml/handler.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+	}
+	keyPair.Leaf, err = x509.ParseCertificate(keyPair.Certificate[0])
+	if err != nil {
+		return err
+	}
+
+	var idpMetadata *samlidp.EntityDescriptor
+
+	// We check if the metadata file is provided
+	if c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"] != "" {
```
Comment (psauvage0) [inline selfservice/flow/saml/handler.go]: OOB ? 
Also, you can check if a map entry is set with something like `if value, ok := mymap["key"]; ok { ...`
```suggestion
	if metadataURL, ok := c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"]; ok {
```

File: selfservice/flow/saml/handler.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+	if err != nil {
+		return err
+	}
+
+	var publicUrlString = config.SelfPublicURL().String()
+
+	// Sometimes there is an issue with double slash into the url so we prevent it
+	// Crewjam library use default route for ACS and metadat but we want to overwrite them
+	RouteSamlAcsWithSlash := RouteSamlAcs
+	if publicUrlString[len(publicUrlString)-1] != '/' {
```
Comment (psauvage0) [inline selfservice/flow/saml/handler.go]: You could probably simplify this with something like `strings.TrimSuffix(strings.TrimSuffix(publicURLString, "/"), "/") + "/"`.
See https://pkg.go.dev/strings#TrimSuffix for more info

File: selfservice/flow/saml/handler.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+	} else if publicUrlString[len(publicUrlString)-1] == '/' {
+
+		publicUrlString = publicUrlString[:len(publicUrlString)-1]
+		u, err := url.Parse(publicUrlString + RouteSamlAcsWithSlash)
+		if err != nil {
+			return err
+		}
+		samlMiddleWare.ServiceProvider.AcsURL = *u
+	}
```
Comment (psauvage0) [inline selfservice/flow/saml/handler.go]:
```suggestion
  u, err := url.Parse(strings.TrimSuffix(strings.TrimSuffix(publicURLString, "/"), "/") + "/"+ RouteSamlAcsWithSlash)
  if err != nil {
	  return err
  }
  samlMiddleWare.ServiceProvider.AcsURL = *u

```

File: selfservice/flow/saml/handler.go [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+	samlMiddleWare.ServiceProvider.AuthnNameIDFormat = samlidp.UnspecifiedNameIDFormat
+
+	samlMiddleware = samlMiddleWare
+
+	return nil
+}
+
+func GetMiddleware() (*samlsp.Middleware, error) {
+	if samlMiddleware == nil {
+		return nil, errors.Errorf("The MiddleWare for SAML is null (Probably due to a backward step)")
```
Comment (psauvage0) [inline selfservice/flow/saml/handler.go]: There is no null in go :p

Comment (psauvage0) [reply selfservice/flow/saml/handler.go]: Joke aside, I don't understand the backward step part, the error message is a bit unclear

File: selfservice/strategy/saml/provider.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/ory/kratos/driver/config"
+	"github.com/ory/kratos/x"
+)
+
+type Provider interface {
+	Claims(ctx context.Context, config *config.Config, SAMLAttribute samlsp.Attributes) (*Claims, error)
+	Config() *Configuration
+}
+
+type Claims struct {
```
Comment (psauvage0) [inline selfservice/strategy/saml/provider.go:17]: I couldn't find anything like this type Claims in the other strategies... Is it used to handle attribute mapping ? If so see previous comment about attribute mapping config

Comment (psauvage0) [inline selfservice/strategy/saml/provider.go:17]: I couldn't find Claims like those in other strategies. Is it linked to attribute mapping ?

File: selfservice/strategy/saml/provider_saml.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+		Name:          SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["firstname"]),
+		LastName:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["lastname"]),
+		Nickname:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["nickname"]),
+		Gender:        SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["gender"]),
+		Birthdate:     SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["birthdate"]),
+		Picture:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["picture"]),
+		Email:         SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["email"]),
+		PhoneNumber:   SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["phone_number"]),
+		EmailVerified: true,
+	}
```
Comment (psauvage0) [inline selfservice/strategy/saml/provider_saml.go]: Attribute mapping needs to be able to take into account user defined attributes. What if I need the attribute `group` from the IDP to map to a `group` trait I configured in a Kratos Identity schema ?

File: selfservice/strategy/saml/strategy/strategy.go [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+			}
+		}
+	}
+
+	return attributes, nil
+}
+
+// Handle /selfservice/methods/saml/acs | Receive SAML response, parse the attributes and start auth flow
+func (s *Strategy) handleCallback(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	r.ParseForm()
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: Error is not checked

File: selfservice/strategy/saml/strategy/strategy.go [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+// Return the SAML Provider
+func (s *Strategy) Provider(ctx context.Context) (samlstrategy.Provider, error) {
+	c, err := s.Config(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	provider, err := c.Provider(c.SAMLProviders[len(c.SAMLProviders)-1].ID, c.SAMLProviders[len(c.SAMLProviders)-1].Label)
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: No OOB possible ?

File: .schema/api.openapi.json [hunk 13]
```diff
@@ -1932,6 +1932,81 @@
         ]
       }
     },
+    "/self-service/saml/metadata":{
+      "get":{
+        "description": "This endpoint is for the IDP to obtain kratos metadata",
```
Comment (Stoakes) [inline .schema/api.openapi.json:1937]: The style for the other descriptions is slightly different.

"This endpoints returns kratos metadata as part of a SAMl login" would be closer to what is done on other endpoints.

File: .schema/api.openapi.json [hunk 14]
```diff
@@ -1932,6 +1932,81 @@
...
+        },
+        "summary": "Expose metadata of the SAML Service Provider (Kratos)",
+        "tags": [
+          "public"
+        ]
+      }
+    },
+    "/self-service/saml/idp":{
+      "get":{
+        "description": "This endpoint is to redirect the user to the idp auth flow",
```
Comment (Stoakes) [inline .schema/api.openapi.json:1962]: Same as above: style for description text is different from other already existing endpoints.

File: selfservice/strategy/saml/strategy/strategy.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+		rf.UI.SetCSRF(s.d.GenerateCSRFToken(r))
+		AddProvider(rf.UI, provider, text.NewInfoRegistrationContinue())
+
+		if traits != nil {
+			ds, err := s.d.Config(r.Context()).DefaultIdentityTraitsSchemaURL()
+			if err != nil {
+				return err
+			}
+
+			traitNodes, err := container.NodesFromJSONSchema(r.Context(), node.OpenIDConnectGroup, ds.String(), "", nil)
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: Use SAMLGroup node ?

File: .schema/api.openapi.json [hunk 16]
```diff
@@ -1932,6 +1932,81 @@
         ]
       }
     },
+    "/self-service/saml/metadata":{
+      "get":{
+        "description": "This endpoint is for the IDP to obtain kratos metadata",
+        "operationId": "getSamlMetadata",
+        "response": {
```
Comment (Stoakes) [inline .schema/api.openapi.json:1939]: I'm missing a few bits on SAML flow, so this might be an irrelevant comment; is it expected there's no 200 response ?

File: selfservice/strategy/saml/strategy/strategy.go [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+			if err != nil {
+				return err
+			}
+
+			rf.UI.Nodes = append(rf.UI.Nodes, traitNodes...)
+			rf.UI.UpdateNodeValuesFromJSON(traits, "traits", node.OpenIDConnectGroup)
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: Here too

File: .schema/openapi.json [hunk 18]
```diff
@@ -2006,6 +2006,81 @@
         ]
       }
     },
+    "/self-service/saml/idp":{
+      "get":{
+        "description": "This endpoint is to redirect the user to the idp auth flow",
```
Comment (Stoakes) [inline .schema/openapi.json:2011]: All other descriptions are something like "This endpoint + verb at the active form".

File: selfservice/strategy/saml/strategy/strategy.go [hunk 19]
```diff
@@ -0,0 +1,352 @@
...
+				return 0, errors.WithStack(err)
+			}
+
+			for _, ider := range c.Identifiers {
+				parts := strings.Split(ider, ":")
+				if len(parts) != 2 {
+					continue
+				}
+
+				if parts[0] == conf.Providers[0].Provider && parts[1] == conf.Providers[0].Subject && len(conf.Providers[0].Subject) > 1 && len(conf.Providers[0].Provider) > 1 {
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: Possible OOB with `conf.Providers[0]` ?

File: embedx/config.schema.json [hunk 20]
```diff
@@ -312,6 +312,225 @@
...
+          "type": "string",
+          "format": "uri",
+          "examples": [
+            "file://path/to/key",
+            "https://foo.bar.com/path/to/key"
+          ]
+        },
+        "mapper_url": {
+          "title": "Jsonnet Mapper URL",
+          "description": "The URL where the jsonnet source is located for mapping the provider's data to Ory Kratos data.",
```
Comment (Stoakes) [inline embedx/config.schema.json:404]: misc english wording

```suggestion
          "description": "Location for the jsonnet mapping between provider's data and Ory Kratos data",
```

File: embedx/config.schema.json [hunk 21]
```diff
@@ -312,6 +312,225 @@
...
+              "type": "string",
+              "examples": [
+                "file://path/to/oidc.jsonnet",
+                "https://foo.bar.com/path/to/oidc.jsonnet",
+                "base64://bG9jYWwgc3ViamVjdCA9I..."
+              ]
+            },
+            "firstname": {
+              "title": "Firstname",
+              "description": "The name of the IDP attribute that will represent the user's firstname on Kratos",
```
Comment (Stoakes) [inline embedx/config.schema.json]: Here and for other attributes

```suggestion
              "description": "Name of the IDP attribute representing user firstname on Kratos",
```

File: selfservice/strategy/saml/strategy/strategy.go [hunk 22]
```diff
@@ -0,0 +1,352 @@
...
+			}
+		}
+	}
+	return
+}
+
+func (s *Strategy) CountActiveFirstFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
+	for _, c := range cc {
+		if c.Type == s.ID() && gjson.ValidBytes(c.Config) {
+			var conf identity.CredentialsOIDC
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: `identity.CredentialsOIDC` in SAML strategy ?

File: persistence/sql/migrations/legacy/20220415152108_credentials_types_saml.up.fizz [hunk 23]
```diff
@@ -0,0 +1 @@
+sql("INSERT INTO identity_credential_types (id, name) SELECT 'ff5a1823-8b47-4255-860f-4b70ed122740', 'saml' WHERE NOT EXISTS ( SELECT * FROM identity_credential_types WHERE name = 'saml')")
```
Comment (Stoakes) [inline persistence/sql/migrations/legacy/20220415152108_credentials_types_saml.up.fizz]: Sorry for missing context: what is behind `ff5a1823-8b47-4255-860f-4b70ed122740` ?

File: selfservice/strategy/saml/strategy/strategy.go [hunk 24]
```diff
@@ -0,0 +1,352 @@
...
+						count++
+					}
+				}
+			}
+		}
+	}
+	return
+}
+
+func (s *Strategy) CountActiveMultiFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy.go]: Not implemented ?

File: postgres.yaml [hunk 25]
```diff
@@ -0,0 +1,33 @@
+version: '3.7'
```
Comment (Stoakes) [inline postgres.yaml:1]: Is this file committed on purpose ?

File: postgres.yaml [hunk 26]
```diff
@@ -0,0 +1,33 @@
...
+
+services:
+  postgresd:
+      image: postgres:9.6
```
Comment (Stoakes) [inline postgres.yaml:5]: This postgres version is antic btw

File: internal/httpclient/model_self_service_saml_url.go [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// SetSamlMetadataUrl sets field value
+func (o *SelfServiceSamlUrl) SetSamlMetadataUrl(v string) {
+	o.SamlMetadataUrl = v
+}
+
+func (o SelfServiceSamlUrl) MarshalJSON() ([]byte, error) {
+	toSerialize := map[string]interface{}{}
+	if true {
```
Comment (Stoakes) [inline internal/httpclient/model_self_service_saml_url.go:95]: This condition will always be true, so it can be removed.

Same for the other ones below.

File: internal/httpclient/model_self_service_saml_url.go [hunk 28]
```diff
@@ -0,0 +1,138 @@
...
+func NewSelfServiceSamlUrlWithDefaults() *SelfServiceSamlUrl {
+	this := SelfServiceSamlUrl{}
+	return &this
+}
+
+// GetSamlAcsUrl returns the SamlAcsUrl field value
+func (o *SelfServiceSamlUrl) GetSamlAcsUrl() string {
+	if o == nil {
+		var ret string
+		return ret
```
Comment (Stoakes) [inline internal/httpclient/model_self_service_saml_url.go:49]: Could be compacted into a single `return ""` statement.

File: internal/httpclient/model_self_service_saml_url.go [hunk 29]
```diff
@@ -0,0 +1,138 @@
...
+// SetSamlAcsUrl sets field value
+func (o *SelfServiceSamlUrl) SetSamlAcsUrl(v string) {
+	o.SamlAcsUrl = v
+}
+
+// GetSamlMetadataUrl returns the SamlMetadataUrl field value
+func (o *SelfServiceSamlUrl) GetSamlMetadataUrl() string {
+	if o == nil {
+		var ret string
+		return ret
```
Comment (Stoakes) [inline internal/httpclient/model_self_service_saml_url.go:73]: Same comment than L49

File: selfservice/flow/saml/handler.go (hunk 5 above)
Comment (Stoakes) [inline selfservice/flow/saml/handler.go]: I feel like this access lacks safety railings:

 * if len(c.SAMLProviders) == 0 this method will panic. Example at https://go.dev/play/p/AAD95wCtX_P
 * Is IDPInformation always initialized ?
 * Is idp_metadata_url always set ? Aren't there cases when it is empty but still valid ?

```suggestion
	if len(c.SAMLProviders) > 0 {
	     if val, ok := c.SAMLProviders[len(c.SAMLProviders)-1].IDPInformation["idp_metadata_url"]; ok && len(val) > 0 {
```

File: selfservice/strategy/saml/strategy/strategy_auth.go [hunk 30]
```diff
@@ -0,0 +1,61 @@
...
+				}
+			}
+
+			return nil, nil
+
+		} else {
+			return nil, err
+		}
+	} else {
+		// The user already exist in database so we log him
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy_auth.go]: This makes me think : what happens when the IDP deletes a user ? I guess the user remains in the SP forever... What are the consequences of that ? It should not be a problem at all, but we could investigate this case a bit

File: selfservice/strategy/saml/provider_config.go [hunk 31]
```diff
@@ -0,0 +1,34 @@
+package saml
+
+type Configuration struct {
+	// ID is the provider's ID
+	ID string `json:"id"`
+
+	// Label represents an optional label which can be used in the UI generation.
+	Label string `json:"label"`
+
+	// Represent the path of the certificate of your application
```
Comment (Stoakes) [inline selfservice/strategy/saml/provider_config.go:18]: Standard Golang comment format is // <variable name> <description>

```suggestion
	// PublicCertPath path of your application certificate
```

Same for other variables below

File: selfservice/strategy/saml/provider_saml.go [hunk 32]
```diff
@@ -0,0 +1,62 @@
...
+	conf := config.SelfServiceStrategy("saml").Config
+	if err := jsonx.
+		NewStrictDecoder(bytes.NewBuffer(conf)).
+		Decode(&c); err != nil {
+		return nil, errors.Wrapf(err, "Unable to decode config %v", string(conf))
+	}
+
+	claims := &Claims{
+		Issuer:        "saml",
+		Subject:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["id"]),
```
Comment (Stoakes) [inline selfservice/strategy/saml/provider_saml.go]: Are we sure at this point of the code that `len(c.SAMLProviders)` is > 0 ?

File: selfservice/strategy/saml/provider_saml.go [hunk 33]
```diff
@@ -0,0 +1,62 @@
...
+		Name:          SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["firstname"]),
+		LastName:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["lastname"]),
+		Nickname:      SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["nickname"]),
+		Gender:        SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["gender"]),
+		Birthdate:     SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["birthdate"]),
+		Picture:       SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["picture"]),
+		Email:         SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["email"]),
+		PhoneNumber:   SAMLAttribute.Get(c.SAMLProviders[len(c.SAMLProviders)-1].AttributesMap["phone_number"]),
+		EmailVerified: true,
```
Comment (Stoakes) [inline selfservice/strategy/saml/provider_saml.go:59]: is `EmailVerified` truly always true at this point of the code ?

File: selfservice/strategy/saml/strategy/strategy.go [hunk 34]
```diff
@@ -0,0 +1,352 @@
...
+
+func (s *Strategy) populateMethod(r *http.Request, c *container.Container, message func(provider string) *text.Message) error {
+	_, err := s.Config(r.Context())
+	if err != nil {
+		return err
+	}
+
+	// does not need sorting because there is only one field
+	c.SetCSRF(s.d.GenerateCSRFToken(r))
+	// AddSamlProviders(c, conf.Providers, message)
```
Comment (Stoakes) [inline selfservice/strategy/saml/strategy/strategy.go]: Commented code can be removed

File: selfservice/strategy/saml/strategy/strategy.go [hunk 35]
```diff
@@ -0,0 +1,352 @@
...
+			rf.UI.UpdateNodeValuesFromJSON(traits, "traits", node.OpenIDConnectGroup)
+		}
+
+		return err
+	}
+
+	return err
+}
+
+func (s *Strategy) CountActiveCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
Comment (Stoakes) [inline selfservice/strategy/saml/strategy/strategy.go]: Personal preference: returning explicitly values instead of using named return argument. I do so, because I find it less error prone.

File: selfservice/strategy/saml/strategy/strategy.go [hunk 36]
```diff
@@ -0,0 +1,352 @@
...
+					count++
+				}
+
+			}
+		}
+	}
+	return
+}
+
+func (s *Strategy) CountActiveFirstFactorCredentials(cc map[identity.CredentialsType]identity.Credentials) (count int, err error) {
```
Comment (Stoakes) [inline selfservice/strategy/saml/strategy/strategy.go]: This function requires a bit of reading to fully understand what it does; it might be interesting to add a comment describing what it does.

File: selfservice/strategy/saml/strategy/strategy_auth.go [hunk 37]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/ory/kratos/selfservice/flow"
+	samlsp "github.com/ory/kratos/selfservice/strategy/saml"
+	"github.com/ory/x/sqlcon"
+)
+
+//#################
+// This file is called when the /ACS receives an assertion, the method below allows you to define whether to register or login the user indicated in the assertion
+//#################
+
+// Handle SAML Assertion and process to either login or register
```
Comment (Stoakes) [inline selfservice/strategy/saml/strategy/strategy_auth.go]: golang standard comment format starts with variable of function name.

Kratos uses a different one: https://github.com/ovh/kratos/blob/af1df63c62a8aec52f6adf9f23bb5dbd38c491ed/selfservice/flow/recovery/handler.go#L203

To be adapted based on your knowledge of ory coding style.

File: selfservice/strategy/saml/strategy/strategy_auth.go [hunk 38]
```diff
@@ -0,0 +1,61 @@
...
+					return nil, s.handleError(w, r, registerFlow, provider.Config().ID, nil, err)
+				} else {
+					return nil, s.handleError(w, r, registerFlow, provider.Config().ID, i.Traits, err)
+				}
+			}
+
+			if err = s.processRegistration(w, r, registerFlow, provider, claims); err != nil {
+				if i == nil {
+					return nil, s.handleError(w, r, registerFlow, provider.Config().ID, nil, err)
+				} else {
```
Comment (Stoakes) [inline selfservice/strategy/saml/strategy/strategy_auth.go]: Unused else: No need to add a `else` when last statement of an `if` is a return.

Here and below

File: selfservice/strategy/saml/strategy/strategy_login.go [hunk 39]
```diff
@@ -0,0 +1,67 @@
...
+	sess.CompletedLoginFor(s.ID(), identity.AuthenticatorAssuranceLevel1) //Add saml to the Authentication Method References
+
+	if err := s.d.LoginHookExecutor().PostLoginHook(w, r, a, i, sess); err != nil {
+		return nil, s.handleError(w, r, a, provider.Config().ID, nil, err)
+	}
+
+	return nil, nil
+}
+
+// Method not used but necessary to implement the interface
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/strategy_login.go]: I find this weird but if you say so 🤨

**Review by Stoakes — COMMENTED** [review]
Nice work !

Some questions and a few comments. Mostly on English, Golang & code style.

File: selfservice/strategy/saml/strategy/test/strategy_test.go [hunk 40]
```diff
@@ -0,0 +1,191 @@
...
+	helpertest "github.com/ory/kratos/selfservice/flow/saml/helpertest"
+	samlstrategy "github.com/ory/kratos/selfservice/strategy/saml/strategy"
+	"github.com/stretchr/testify/require"
+
+	"gotest.tools/assert"
+	gotest "gotest.tools/assert"
+	is "gotest.tools/assert/cmp"
+)
+
+func TestGetAndDecryptAssertion(t *testing.T) {
```
Comment (psauvage0) [inline selfservice/strategy/saml/strategy/test/strategy_test.go]: Unrelated to the code : it would be interesting to discuss test strategy with the Ory community. It seems there is not a lot of info on what is considered a good level of testing on `CONTRIBUING.md` file (should we cover all branches of a functions ? Should we test for edge cases ? Or simply testing the main intent of the function is OK ?)

File: spec/api.json [hunk 41]
```diff
@@ -3636,6 +3653,48 @@
         ]
       }
     },
+    "/self-service/methods/saml/auth": {
```
Comment (psauvage0) [inline spec/api.json:4957]: Same thing as before about the route : `/browser` vs `/api` routes

**Review by psauvage0 — CHANGES_REQUESTED** [review]
Great work ! :D 

The main problem I see is with the attribute mapping that only works with a fixed list of attributes names. 

Otherwise it's about small code quality improvements. There are some pointed out by the linter that I didn't bother to comment as they are already shown in the PR builds.

I honestly think that we could already take reviews from the Ory team & community, as there are maybe some things about how they want the code to be organized.

File: internal/httpclient/api/openapi.yaml (hunk 3 above)
Comment (sebferrer) [reply internal/httpclient/api/openapi.yaml]: We did it on purpose based on a misunderstanding, indeed we should change it back

File: selfservice/strategy/saml/strategy/strategy.go (hunk 24 above)
Comment (sebferrer) [reply selfservice/strategy/saml/strategy/strategy.go]: Not yet. This implementation isn't required since a long time, so we had to do it like this on our last rebase.

File: selfservice/strategy/saml/strategy/strategy_login.go (hunk 39 above)
Comment (sebferrer) [reply selfservice/strategy/saml/strategy/strategy_login.go]: Here's the kind of error we get otherwise:
```
# github.com/ory/kratos/selfservice/strategy/saml/strategy
selfservice/strategy/saml/strategy/strategy_login.go:25:24: cannot use new(Strategy) (value of type *Strategy) as type login.Strategy in variable declaration:
        *Strategy does not implement login.Strategy (missing Login method)
```

Comment (psauvage0) [reply selfservice/strategy/saml/strategy/strategy_login.go]: What I found weird was that the method is called "Login", so I would expect that it is a pretty important method to implement in an IAM. 

Edit: As we discussed, it's because of the rebase against the origin. It's a new method added to the interface, probably worth checking where this method is called, to ensure that it is safe to not implement it

File: embedx/config.schema.json (hunk 2 above)
Comment (ThibHrrd) [reply embedx/config.schema.json:493]: For the moment, this part is not configurable, that is to say that the identity cannot have more fields than what is present here.

File: selfservice/strategy/saml/provider_saml.go (hunk 10 above)
Comment (ThibHrrd) [reply selfservice/strategy/saml/provider_saml.go]: For the moment, the user can't add attributes as he wants but maybe I should add a group attribute

File: selfservice/strategy/saml/strategy/strategy_auth.go (hunk 30 above)
Comment (ThibHrrd) [reply selfservice/strategy/saml/strategy/strategy_auth.go]: This is indeed what will happen but it is complicated for the SP to know that the IDP has deleted a user
